airflow/dag/serialization.py

```diff
@@ -0,0 +1,250 @@
+        if k in x:
+            setattr(new_x, k, _deserialize(x[k], visited_dags))
+        else:
+            setattr(new_x, k, None)
+
+
+def _deserialize_dag(x, visited_dags):
+    """Deserialize a DAG."""
+    if isinstance(x, dict):
+        dag = models.DAG(dag_id=x['_dag_id'])
+        visited_dags[dag.dag_id] = dag
+        _copy_attr(x, dag, _dag_included_fields, visited_dags)
+        dag.is_stringified = True
+        return dag
+    elif isinstance(x, str):
+        if x in visited_dags:
+            return visited_dags[x]
+        logging.warning('DAG %s not found in serialization.', x)
+    else:
+        logging.warning('Invalid type %s for decoded DAG.', type(x))
+    return FAILED
+
+
+def _deserialize_operator(x, operator_class, visited_dags):
+    """Deserialize an operator."""
+    op = object.__new__(models.BaseOperator)
+    _copy_attr(x, op, _op_included_fields, visited_dags)
+    op.task_type = operator_class
+    op.is_stringified = True
+    return op
+
+
+def _deserialize(encoded_x, visited_dags):  # pylint: disable=too-many-return-statements
+    """Helper function of depth first search for deserialization."""
+    try:
+        # JSON primitives (except for dict) are not encoded.
+        if _is_primitive(encoded_x):
+            return encoded_x
+        elif isinstance(encoded_x, list):
+            return [_deserialize(v, visited_dags) for v in encoded_x]
+
+        assert isinstance(encoded_x, dict)
+        var = encoded_x[VAR]
+        type_ = encoded_x[TYPE]
+
+        if type_ == DICT:
+            return {k: _deserialize(v, visited_dags) for k, v in var.items()}
+        elif type_ == DAG:
+            return _deserialize_dag(var, visited_dags)
+        elif type_ == OP:
+            return _deserialize_operator(
+                var,
+                encoded_x[CLASS] if CLASS in encoded_x else BASE_OPERATOR_CLASS,
+                visited_dags)
+        elif type_ == DATETIME:
+            return dateutil.parser.parse(var)
+        elif type_ == TIMEDELTA:
+            return datetime.timedelta(seconds=var)
+        elif type_ == TIMEZONE:
+            return pendulum.tz.timezone(name=var)
+        elif type_ == SET:
+            return {_deserialize(v, visited_dags) for v in var}
+        elif type_ == TUPLE:
+            return tuple([_deserialize(v, visited_dags) for v in var])
+        else:
+            logging.warning('Invalid type %s in deserialization.', type_)
+            return FAILED
+    except Exception:  # pylint: disable=broad-except
+        logging.warning('Failed to deserialize %s.', encoded_x, exc_info=True)
+        return FAILED
+
+
+def serialize(x):
+    """Stringifies DAGs and operators contained by x and returns a JSON string of x."""
+    return json.dumps(_serialize(x, {}))
+
+
+def deserialize(encoded_x):
+    """Deserializes encoded_x and reconstructs all DAGs and operators it contains."""
+    # Setting strict to False for control characters ('\n') in function source code.
```

Comment:
Rather than doing that is it not better to properly JSON escape the source code?

Reply:
Done. Used ensure_ascii=True in json.dumps instead.